contrib/log_correlation/log4j2/src/main/java/io/opencensus/contrib/logcorrelation/log4j2/ContextDataUtils.java

```diff
@@ -0,0 +1,230 @@
+        if (spanContext.getTraceOptions().isSampled()) {
+          return spanContext;
+        } else {
+          return null;
+        }
+      case ALL_SPANS:
+        return getCurrentSpanContext();
+    }
+    throw new AssertionError("Unknown spanSelection: " + spanSelection);
+  }
+
+  private static StringMap getShareableContextData() {
+    ReadOnlyThreadContextMap context = ThreadContext.getThreadContextMap();
+
+    // Return a new object, since StringMap is modifiable.
+    return context == null
+        ? new SortedArrayStringMap(ThreadContext.getImmutableContext())
+        : new SortedArrayStringMap(context.getReadOnlyContextData());
+  }
+
+  private static ReadOnlyStringMap getNonShareableContextData() {
+    ReadOnlyThreadContextMap context = ThreadContext.getThreadContextMap();
+    if (context != null) {
+      return context.getReadOnlyContextData();
+    } else {
+      Map<String, String> contextMap = ThreadContext.getImmutableContext();
+      return contextMap.isEmpty()
+          ? UnmodifiableReadOnlyStringMap.EMPTY
+          : new UnmodifiableReadOnlyStringMap(contextMap);
+    }
+  }
+
+  private static StringMap getShareableContextAndTracingData(SpanContext spanContext) {
+    ReadOnlyThreadContextMap context = ThreadContext.getThreadContextMap();
+    SortedArrayStringMap stringMap =
+        context == null
+            ? new SortedArrayStringMap(ThreadContext.getImmutableContext())
+            : new SortedArrayStringMap(context.getReadOnlyContextData());
+    stringMap.putValue(
+        OpenCensusTraceContextDataInjector.TRACE_ID_CONTEXT_KEY,
+        new TraceIdToLowerBase16Formatter(spanContext));
+    stringMap.putValue(
+        OpenCensusTraceContextDataInjector.SPAN_ID_CONTEXT_KEY,
+        new SpanIdToLowerBase16Formatter(spanContext));
+    stringMap.putValue(
+        OpenCensusTraceContextDataInjector.TRACE_SAMPLED_CONTEXT_KEY,
+        spanContext.getTraceOptions().isSampled() ? "true" : "false");
+    return stringMap;
+  }
+
+  private static SpanContext getCurrentSpanContext() {
+    Span span = ContextUtils.CONTEXT_SPAN_KEY.get();
+    return span == null ? SpanContext.INVALID : span.getContext();
+  }
+
+  private static final class TraceIdToLowerBase16Formatter {
+    private final SpanContext spanContext;
+
+    private TraceIdToLowerBase16Formatter(SpanContext spanContext) {
+      this.spanContext = spanContext;
+    }
+
+    @Override
+    public String toString() {
+      return spanContext.getTraceId().toLowerBase16();
+    }
+  }
+
+  private static final class SpanIdToLowerBase16Formatter {
+    private final SpanContext spanContext;
+
+    private SpanIdToLowerBase16Formatter(SpanContext spanContext) {
+      this.spanContext = spanContext;
+    }
+
+    @Override
+    public String toString() {
+      return spanContext.getSpanId().toLowerBase16();
+    }
+  }
+
+  @Immutable
+  private static final class UnmodifiableReadOnlyStringMap implements ReadOnlyStringMap {
```

Comment:
nit: I think you can use SortedArrayStringMap + freeze() to get the same behavior and not have to implement this. Motivation is that they to some special thing when copy from a SortedArrayStringMap (see constructor or putAll).